code/go/0chain.net/smartcontract/storagesc/challenge.go

```diff
@@ -972,57 +974,82 @@ func (sc *StorageSmartContract) populateGenerateChallenge(
 		return nil, errors.New("invalid blobber for allocation")
 	}
 
+	if alloc.DataShards+1 > needValidNum {
+		needValidNum = alloc.DataShards + 1
+	}
+
 	var randValidators []ValidationPartitionNode
 	if err := validators.GetRandomItems(balances, r, &randValidators); err != nil {
 		return nil, common.NewError("add_challenge",
 			"error getting validators random slice: "+err.Error())
 	}
 
+	if len(randValidators) < needValidNum {
+		return nil, errors.New("validators number does not meet minimum challenge requirement")
+	}
+
 	var (
-		needValidNum       = minInt(len(randValidators), alloc.DataShards+1)
-		selectedValidators = make([]*ValidationNode, 0, needValidNum)
-		perm               = r.Perm(len(randValidators))
+		selectedValidators  = make([]*ValidationNode, 0, needValidNum)
+		perm                = r.Perm(len(randValidators))
+		remainingValidators = len(randValidators)
 	)
 
 	now := txn.CreationDate
 	filterValidator := filterHealthyValidators(now)
 
-	for i := 0; i < needValidNum; i++ {
+	for i := 0; i < len(randValidators) && len(selectedValidators) < needValidNum; i++ {
+		if remainingValidators < needValidNum {
+			return nil, errors.New("validators number does not meet minimum challenge requirement after filtering")
+		}
 		randValidator := randValidators[perm[i]]
-		if randValidator.Id != blobberID {
-			validator, err := getValidator(randValidator.Id, balances)
-
-			if err != nil {
-				if cstate.ErrInvalidState(err) {
-					return nil, common.NewError("add_challenge",
-						err.Error())
-				}
-				continue
-			}
-
-			kick, err := filterValidator(validator)
-
-			if err != nil {
-				return nil, common.NewError("add_challenge", "failed to filter validator: "+
+		if randValidator.Id == blobberID {
+			continue
+		}
+		validator, err := getValidator(randValidator.Id, balances)
+		if err != nil {
+			if cstate.ErrInvalidState(err) {
+				return nil, common.NewError("add_challenge",
 					err.Error())
 			}
+			continue
+		}
 
-			if kick {
-				continue
-			}
-
-			selectedValidators = append(selectedValidators,
-				&ValidationNode{
-					Provider: provider.Provider{
-						ID:           randValidator.Id,
-						ProviderType: spenum.Validator,
-					},
-					BaseURL: randValidator.Url,
-				})
+		kick, err := filterValidator(validator)
+		if err != nil {
+			return nil, common.NewError("add_challenge", "failed to filter validator: "+
+				err.Error())
 		}
-		if len(selectedValidators) > alloc.DataShards {
-			break
+		if kick {
+			remainingValidators--
+			continue
 		}
+
+		// sp, err := sc.getStakePool(spenum.Validator, validator.ID, balances)
+		// if err != nil {
+		// 	return nil, fmt.Errorf("can't get validator %s stake pool: %v", randValidator.Id, err)
+		// }
+		// stake, err := sp.stake()
+		// if err != nil {
+		// 	return nil, err
+		// }
+		// if stake < minStake {
```

Comment:
@din-mukhammed any reason for removing the min stake checking.

Reply:
Oh, it was not needed originally, if so, please remove the commented code and remove the `minStake` param, thanks.

Reply:
@peterlimg I was just testing to see if this min stake check was the cause of failing system tests. The check was requested from this [issue](https://github.com/0chain/0chain/issues/587), can you please confirm if it's the right way to filter out the validators with stake lower than minimum during `populateGenerateChallenge`? Anyway I can create a separate PR for that issue and unblock this one.

Reply:
@din-mukhammed yes, let's do it in another PR.